### PR TITLE
Add authentication support for default registries

### PR DESCRIPTION
**Pull request by runcom**
https://trello.com/c/TXe7uAER/36-5-back-port-1-9-x-add-authentication-support-for-default-registries

@rhatdan @miminar

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by runcom** [comment]
@miminar taking care of https://github.com/runcom/docker/commit/52f9224edda45e1017dad3f10ec610e8dcea3d34#commitcomment-15338406

**Comment by rhatdan** [comment]
@miminar PTAL

**Comment by miminar** [comment]
Already commented in [WIP branch](https://github.com/runcom/docker/commit/52f9224edda45e1017dad3f10ec610e8dcea3d34). It's being addressed.

**Comment by runcom** [comment]
@miminar took care of your comment, now it should be good to go (except what I was talking about wrt `docker push busybox` failing because of auth, which the trello card doesn't cover so I guess it's ok)

File: api/server/router/local/image.go [hunk 1]
```diff
@@ -584,25 +635,49 @@ func (s *router) getImagesSearch(ctx context.Context, w http.ResponseWriter, r *
...
+
+	// this is the case when we fully qualify images
+	repoInfo, err := registry.ParseRepositoryInfo(term)
+	if err != nil {
+		return err
+	}
+	if len(configs) == 0 {
+		configs[repoInfo.Index.Name] = authConfig
+	}
+
```
Comment (miminar) [inline api/server/router/local/image.go]: There's too much rendundancy in this file. Could you put the authparsing into a helper?

Comment (runcom) [inline api/server/router/local/image.go]: Yep, noticed it as well, I'm adding tests right now, I'll fix this along with them asap

**Comment by runcom** [comment]
@miminar added tests and deduplicate code

File: integration-cli/docker_cli_pull_test.go [hunk 2]
```diff
@@ -757,3 +758,34 @@ func (s *DockerRegistriesSuite) TestPullFromBlockedRegistry(c *check.C) {
...
+
+	// push the image so we can pull from unqualified "runcom/busybox"
+	out, err = s.d.Cmd("push", repo)
+	c.Assert(err, check.IsNil, check.Commentf(out))
+
+	out, err = s.d.Cmd("pull", repoUnqualified)
+	c.Assert(err, check.IsNil, check.Commentf(out))
+	// check it's pulling from private reg with auth with unqualified images
+	expected := fmt.Sprintf("Trying to pull repository %s ... latest: Pulling from %s", repo, repoUnqualified)
+	if !strings.Contains(out, expected) {
```
Comment (miminar) [inline integration-cli/docker_cli_pull_test.go]: Please confirm that no attempt to pull from `docker.io` is done. Or that resulting image is prefixed with expected repo name.

File: integration-cli/docker_cli_pull_test.go [hunk 3]
```diff
@@ -757,3 +758,34 @@ func (s *DockerRegistriesSuite) TestPullFromBlockedRegistry(c *check.C) {
...
+	out, err := s.d.Cmd("tag", "busybox", repo)
+	c.Assert(err, check.IsNil, check.Commentf(out))
+
+	// this means it needs auth...
+	resp, err := http.Get(fmt.Sprintf("http://%s/v2/", s.regWithAuth.url))
+	c.Assert(err, check.IsNil)
+	c.Assert(resp.StatusCode, check.Equals, http.StatusUnauthorized)
+
+	// login with the registry...
+	out, err = s.d.Cmd("login", "-u", "testuser", "-p", "testpassword", "-e", "test@test.org", s.regWithAuth.url)
```
Comment (miminar) [inline integration-cli/docker_cli_pull_test.go]: I'd make `testuser` and `testpassword` attributes of the `regWithAuth` registry.

File: integration-cli/docker_cli_push_test.go [hunk 4]
```diff
@@ -554,3 +554,41 @@ func (s *DockerRegistrySuite) TestPushCustomTagToAdditionalRegistry(c *check.C)
...
+	repo := fmt.Sprintf("%s/runcom/busybox", s.regWithAuth.url)
+	repoUnqualified := "runcom/busybox"
+
+	out, err := s.d.Cmd("tag", "busybox", repoUnqualified)
+	c.Assert(err, check.IsNil, check.Commentf(out))
+
+	// try to pull with unqualified image and get auth denied
+	out, err = s.d.Cmd("push", repoUnqualified)
+	c.Assert(err, check.NotNil, check.Commentf(out))
+	expected := "Post http://127.0.0.1:5002/v2/runcom/busybox/blobs/uploads/: no basic auth credentials"
```
Comment (miminar) [inline integration-cli/docker_cli_push_test.go]: Use `fmt.Sprintf()` to expand `s.regWithAuth.url` and repo name.

File: integration-cli/registry.go [hunk 5]
```diff
@@ -20,24 +20,38 @@ type testRegistryV2 struct {
...
-	tmp, err := ioutil.TempDir("", "registry-test-")
-	if err != nil {
-		return nil, err
+    addr: %s
+%s`
+	var htpasswd string
+	if auth {
+		htpasswdPath := filepath.Join(tmp, "htpasswd")
+		// testuser:testpassword
+		if err := ioutil.WriteFile(htpasswdPath, []byte(`testuser:$2y$05$sBsSqk0OpSD1uTZkHXc4FeJ0Z70wLQdAX/82UiHuQOKbNbBrzs63m`), os.FileMode(0644)); err != nil {
```
Comment (miminar) [inline integration-cli/registry.go]: Add at least a comment, how `$2y$05$sBsSq...` can be created.

**Comment by miminar** [comment]
Few nitpicks from my side. Great work otherwise!

**Comment by miminar** [comment]
LGTM

**Comment by runcom** [comment]
@rhatdan as agreed with @miminar, I'm finishing to fix all the tests and let you know when this will be good to merge

**Comment by rhatdan** [comment]
Ok ping me when it is done, and I will do a quick scan.

**Comment by runcom** [comment]
@rhatdan everything's good now, you can review and merge now

**Comment by rhatdan** [comment]
LGTM Except for the nasty size of this patch.  Are we merging something like this into docker-1.10 also?

**Comment by miminar** [comment]
This patch should be merged into registry patches. It affects at least 3 of them.

It needs to be applied to docker-1.10 branch as well once the registry patches are rebased.

**Comment by rhatdan** [comment]
That is what I thought.  @runcom has another patch that should also be merged docker info lists the registries that were added, I believe.

**Comment by runcom** [comment]
@rhatdan yes PR is at https://github.com/projectatomic/docker/pull/32

**Comment by miminar** [comment]
I'll merge the info patch.

**Pull request merged**